daffodil-test/src/test/scala/org/apache/daffodil/section00/general/TestDisallowDocType.scala

```diff
@@ -87,7 +86,7 @@ class TestDisallowDocType {
   }
 
   @Test def test_infosetFileMustNotHaveDocType(): Unit = {
-    val e = intercept[TDMLException] {
+    val e = intercept[SAXParseException] {
```

Comment:
Seems we shouldn't let SAXParseExceptions bubble  to the very top? I think the way the DaffodilConstructorLoader is intended to work is it doesn't throw exceptions but instead gathers up all errors in the error handler diagnostics to we can report them later?

Reply:
It still needs to stop execution on the fatal error here or it will get caught in the infinite loop. I originally had fatalError throwing an SDE but then I found out the TDML runner installs its own custom error handler. So, to mimic the same behavior I tried to throw an SDE in that runner but it appears that the concrete implementations of those are in runtime1 and not lib. That is why I left it as a SAXParseException. Should the TDML runner catch any unhanded SAXParseExceptions and wrap them in TDMLExceptions or is there a better way?

Reply:
>  Should the TDML runner catch any unhanded SAXParseExceptions and wrap them in TDMLExceptions or is there a better way?

It probably depends where the errors are coming from. If the errors are from compiling a DFDL schema used by a TDML test, those error should all end up as `Diagnositc`s that the TDML runner can look for. If they are coming from something else (e.g. TDMLRunner validating a .tdml file), then it probably wants to be caught and turned into a TDML exception.

Reply:
I wonder if the problem of the leaking exception is here:

https://github.com/apache/daffodil/blob/b1a5ab06c3bd194ecd43f37789654c89c83beb49/daffodil-lib/src/main/scala/org/apache/daffodil/lib/xml/DaffodilConstructingLoader.scala#L477

When parsing fatalError will be called which throws an exception. That is likely caught in this catch block, but then calls fatalError which will throw that exception. Not only does this mean we'll see the exception twice, it means the exception leaks out. It probably also means the exception is duplicated in the diagnostics, unless the exception was thrown by scala-xml without calling fatalError.

Reply:
I found the root cause of the issue. Daffodil was working fine. It is a bug in scala-xml. In this specific case, the infinite loop is caused when the scanner does not advance when this error is reported. There is evidence of error correction elsewhere so I think this was simply overlooked. My proposed solution is to now only throw an exception when that specific error is raised. Yes, looking at the error message is a bit hackish, but I'm thinking that it is a good compromise where it can continue as much as possible without causing the infinite loop. We have a test now and if the message changes, it should be caught.

Reply:
Do you think the fix in scala-xml would be fairly straightforward? It would be nice if we could submit a PR so we can eventually get rid of the special casing, since I agree it does sound like a bit of a hack.

Reply:
The code for that is as follows:

```scala
  /**
   * attribute value, terminated by either `'` or `"`. value may not contain `<`.
   * @param endCh either `'` or `"`
   */
  def xAttributeValue(endCh: Char): String = {
    val buf = new StringBuilder
    while (ch != endCh && !eof) {
      // well-formedness constraint
      if (ch == '<') reportSyntaxError("'<' not allowed in attrib value")
      else if (ch == SU) truncatedError("")
      else buf append ch_returning_nextch
    }
    ch_returning_nextch
    // @todo: normalize attribute value
    buf.toString
  }
```

I think that when a `<` character is seen, the error recovery needs to be a call to `ch_returning_nextch`  without appending it to `buf`. So I think it is probably straightforward but I'm not sure. The call to `truncatedError` doesn't seem to handle that either. Maybe there is some other expectation here. It might be worth time boxing a fix attempt. If a working PR can be prepared in less than X hours, great, otherwise drop the attempt.

Reply:
Do we want to keep this explicit check in for now and remove it if, and when, a fix gets merged into scala-xml? I'll try to get a fix in over there but there is no guarantee on when that could happen.